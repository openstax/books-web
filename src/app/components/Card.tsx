import styled from 'styled-components';
import { textStyle } from './Typography/base';

const margin = 3.0;

// tslint:disable-next-line:variable-name
const Card = styled.div`
  display: flex;
  flex-direction: column;
  margin: auto;
  overflow: hidden;
  padding: ${margin}rem;
  width: 30rem;
  ${textStyle};
  background-color: white;
  box-shadow: 0 0 2rem rgba(0, 0, 0, 0.05), 0 0 4rem rgba(0, 0, 0, 0.08);
  border-radius: 0.5rem;
`;

Card.Header = styled.header`
  display: flex;
  align-items: center;
  margin-bottom: ${margin * 0.5}rem;
`;

Card.Heading = styled.h1`
  display: flex;
  align-items: center;
  margin: 0;
  font-size: 2.4rem;
  font-weight: bold;
  text-align: center;
`;

Card.Body = styled.div`
  display: flex;
  flex-direction: column;
  font-size: 1.4rem;
  margin-top: ${margin * 0.5}rem;
  margin-bottom: ${margin * 0.5}rem;

  ${Card.Header} + & { /* stylelint-disable */
    margin-top: 0;
  }
`;

Card.Footer = styled.div`
  display: flex;
  justify-content: flex-end;
`;

export default Card;
