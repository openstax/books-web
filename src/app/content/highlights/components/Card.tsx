import { Highlight } from '@openstax/highlighter';
import { HTMLElement } from '@openstax/types/lib.dom';
import flow from 'lodash/fp/flow';
import React from 'react';
import { connect } from 'react-redux';
import styled, { css } from 'styled-components/macro';
import { findElementSelfOrParent, scrollIntoView } from '../../../domUtils';
import theme from '../../../theme';
import { AppState, Dispatch } from '../../../types';
import { assertWindow, remsToEms } from '../../../utils';
import { contentTextWidth, searchResultsBarDesktopWidth, sidebarDesktopWidth } from '../../components/constants';
import { disablePrint } from '../../components/utils/disablePrint';
import { styleWhenSidebarClosed } from '../../components/utils/sidebar';
import * as selectHighlights from '../../highlights/selectors';
import * as selectSearch from '../../search/selectors';
import * as contentSelect from '../../selectors';
import { clearFocusedHighlight, createHighlight, deleteHighlight, updateHighlight } from '../actions';
import {
  cardContentMargin,
  cardFocusedContentMargin,
  cardMinWindowMargin,
  cardPadding,
  cardWidth,
  highlightStyles
} from '../constants';
import { HighlightData } from '../types';
import DisplayNote from './DisplayNote';
import EditCard from './EditCard';
import { cardBorder } from './style';

interface Props {
  container: HTMLElement;
  isFocused: boolean;
  highlight: Highlight;
  create: typeof createHighlight;
  save: typeof updateHighlight;
  remove: typeof deleteHighlight;
  blur: typeof clearFocusedHighlight;
  data?: HighlightData;
  className: string;
}

// tslint:disable-next-line:variable-name
const Card = (props: Props) => {
  const note = props.data && props.data.note;
  const element = React.useRef<HTMLElement>(null);
  const [editing, setEditing] = React.useState<boolean>(!note);

  React.useEffect(() => {
    if (element.current && props.isFocused) {
      scrollIntoView(element.current);
    }
    if (!props.isFocused) {
      setEditing(false);
    }
  }, [props.isFocused]);

  if (!props.highlight.range) {
    return null;
  }

  const onRemove = () => props.data && props.remove(props.data.id);
  const style = highlightStyles.find((search) => props.data && search.label === props.data.style);

  const commonProps = {
    className: props.className,
    isFocused: props.isFocused,
    onBlur: props.blur,
    onRemove,
    ref: element,
  };

  return !editing && style && note ? <DisplayNote
    {...commonProps}
    style={style}
    note={note}
    onEdit={() => setEditing(true)}
  /> : <EditCard
    {...commonProps}
    highlight={props.highlight}
    onCreate={props.create}
    onSave={props.save}
    data={props.data}
  />;
};

/*
 * putting overflow hidden on a page wrapper that aligns with the window edge would
 * include the sidebar, which would break position: sticky.
 *
 * avoiding using an overflow hidden to hide cards when there is not enough space
 * means being very explicit about hiding them so they don't create a horizontal
 * scrollbar.
 *
 * in this case that means using extensive knowledge about the container widths,
 * which unfortunately means knowledge of all the sidebar widths and their state
 * too.
 *
 * consider making a helper like `styleWhenSidebarClosed` maybe `styleWhenContentWidth`
 * that has a selector to get the relevant stuff.
 */

const additionalWidthForCard = (cardWidth + cardContentMargin + cardMinWindowMargin) * 2;

const getHighlightOffset = (container: HTMLElement, highlight: Highlight) => {
  if (!highlight.range || !highlight.range.getBoundingClientRect) {
    return;
  }

  const anchor = findElementSelfOrParent(highlight.range.commonAncestorContainer);

  if (!anchor) {
    return;
  }

  const {top, bottom} = highlight.range.getBoundingClientRect();

  const offsetParent = container.offsetParent && findElementSelfOrParent(container.offsetParent);
  const parentOffset = offsetParent ? offsetParent.offsetTop : 0;
  const scrollOffset = assertWindow().scrollY;

  return {
    bottom: bottom - parentOffset + scrollOffset,
    top: top - parentOffset + scrollOffset,
  };
};

const getHighlightTopOffset = (container: HTMLElement, highlight: Highlight): number | undefined => {
  const offset = getHighlightOffset(container, highlight);

  if (offset) {
    return offset.top;
  }
};
const getHighlightBottomOffset = (container: HTMLElement, highlight: Highlight): number | undefined => {
  const offset = getHighlightOffset(container, highlight);

  if (offset) {
    return offset.bottom;
  }
};

const overlapDisplay = css`
  ${(props: Props) => !!props.isFocused && css`
    left: unset;
    right: ${cardMinWindowMargin}rem;
    top: ${() => {
      return getHighlightBottomOffset(props.container, props.highlight) || 0;
    }}px;
  `}
  ${(props: Props) => !props.isFocused && css`
    display: none;
  `}
`;

const rightSideDisplay = css`
  left: calc(100% - ((100% - ${contentTextWidth}rem) / 2) + ${cardContentMargin}rem);
  right: unset;
  top: ${(props: Props) => {
    return getHighlightTopOffset(props.container, props.highlight) || 0;
  }}px;
  ${(props: Props) => !!props.isFocused && css`
    left: calc(100% - ((100% - ${contentTextWidth}rem) / 2) + ${cardFocusedContentMargin}rem);
  `}
  ${(props: Props) => !props.isFocused && css`
    /* temporary simplification */
    display: none;
  `}
`;

const mobileDisplay = css`
  ${(props: Props) => !!props.isFocused && css`
    left: 0;
    right: 0;
    bottom: 0;
    top: unset;
    position: fixed;
    padding: 0;
  `}
  ${(props: Props) => !props.isFocused && css`
    display: none;
  `}
`;

// tslint:disable-next-line:variable-name
const StyledCard = styled(Card)`
  position: absolute;
  padding: ${cardPadding}rem;
  ${cardBorder}
  ${rightSideDisplay}
  ${disablePrint}

  ${(props: {data: HighlightData}) => {
    const data = props.data;

    if (!data || !data.style) {
      return null;
    }

    const style = highlightStyles.find((search) => search.label === props.data.style);

    if (!style) {
      return null;
    }

    return css`
      ::before {
        content: ' ';
        border-radius: 0.4rem 0 0 0.4rem;
        position: absolute;
        top: 0;
        left: 0
        bottom: 0;
        width: ${cardPadding / 2}rem;
        background-color: ${style.focused};
      }
      ${theme.breakpoints.mobile(css`
        ::before {
          border-radius: 0.4rem 0.4rem 0 0;
          right: 0;
          bottom: unset;
          width: unset;
          height: ${cardPadding / 2}rem;
        }
     `)}
    `;
  }}

  @media (max-width: ${remsToEms(contentTextWidth + sidebarDesktopWidth + additionalWidthForCard)}em) {
    /* the window is too small to show note cards next to content when the toc is open */
    ${overlapDisplay}
    ${styleWhenSidebarClosed(rightSideDisplay)}
  }

  ${(props: {hasQuery: boolean}) => !!props.hasQuery && css`
    @media (max-width: ${remsToEms(contentTextWidth + searchResultsBarDesktopWidth + additionalWidthForCard)}em) {
      /* the window is too small to show note cards next to content when search is open */
      ${overlapDisplay}
    }
  `}

  @media (max-width: ${remsToEms(contentTextWidth + additionalWidthForCard)}em) {
    /* the window is too small to show note cards next to content even without sidebars */
    ${overlapDisplay}
  }

  ${theme.breakpoints.mobile(css`
    ${mobileDisplay}
 `)}
`;

export default connect(
  (state: AppState, ownProps: {highlight: Highlight}) => ({
    data: selectHighlights.highlights(state).find((search) => search.id === ownProps.highlight.id),
    hasQuery: !!selectSearch.query(state),
    isFocused: selectHighlights.focused(state) === ownProps.highlight.id,
    isOpen: contentSelect.tocOpen(state),
  }),
  (dispatch: Dispatch) => ({
    blur: flow(clearFocusedHighlight, dispatch),
    create: flow(createHighlight, dispatch),
    remove: flow(deleteHighlight, dispatch),
    save: flow(updateHighlight, dispatch),
  })
)(StyledCard);
