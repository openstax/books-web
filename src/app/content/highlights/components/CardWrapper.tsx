import Highlighter, { Highlight } from '@openstax/highlighter';
import { HTMLElement } from '@openstax/types/lib.dom';
import React from 'react';
import styled from 'styled-components';
import theme from '../../../theme';
import { assertDefined } from '../../../utils';
import { cardMarginBottom } from '../constants';
import Card, { getHighlightTopOffset } from './Card';

interface Props {
  container: HTMLElement;
  highlighter: Highlighter;
  highlights: Highlight[];
  className?: string;
}

// tslint:disable-next-line:variable-name
const Wrapper = ({highlights, className, container, highlighter}: Props) => {
  const element = React.useRef<HTMLElement>(null);
  const [cardsPositions, setCardsPositions] = React.useState<Map<string, number>>(new Map());
  const [cardsHeights, setCardsHeights] = React.useState<Map<string, number>>(new Map());

  const onHeightChange = (id: string, height: number) => {
    if (cardsHeights.get(id) !== height) {
      setCardsHeights((data) => new Map(data.set(id, height)));
    }
  };

  const onFocus = (id: string) => {
    const highlight = highlights.find((search) => search.id === id);
    const position = cardsPositions.get(id);
    if (typeof position !== 'number' || !highlight) { return; }

    const topOffset = assertDefined(
      getHighlightTopOffset(container, highlight),
      `Couldn't get top offset for highlights`
    );

    if (position > topOffset) {
      element.current!.style.top = `-${position - topOffset}px`;
    }
  };

  const onBlur = () => {
    element.current!.style.top = '0';
  };

  const updatePositions = React.useCallback(() => {
    const newPositions: Map<string, number> = new Map();

    let lastVisibleHighlightPosition: number = 0;
    let lastVisibleHighlightHeight: number = 0;

    for (const [index, highlight] of highlights.entries()) {
      const topOffset = assertDefined(
        getHighlightTopOffset(container, highlight),
        `Couldn't get top offset for highlights`
      );

      let stackedTopOffset = lastVisibleHighlightPosition;

      if ((topOffset - lastVisibleHighlightPosition) < stackedTopOffset) {
        stackedTopOffset = stackedTopOffset
          + lastVisibleHighlightHeight
          + (index > 0 ? (cardMarginBottom * 10) : 0); // * 10 because constants are in pixels and we need rems.
      } else {
        stackedTopOffset = topOffset;
      }

      if (cardsHeights.get(highlight.id)) {
        lastVisibleHighlightPosition = stackedTopOffset;
        lastVisibleHighlightHeight = cardsHeights.get(highlight.id)!;
      }

      newPositions.set(highlight.id, stackedTopOffset);

      setCardsPositions(newPositions);
    }
  }, [highlights, cardsHeights, container]);

  React.useEffect(() => {
    updatePositions();
  }, [updatePositions]);

  return <div className={className} ref={element}>
    {highlights.map((highlight) => <Card
      highlighter={highlighter}
      highlight={highlight}
      key={highlight.id}
      container={container}
      topOffset={cardsPositions.get(highlight.id) || 0}
      onHeightChange={onHeightChange}
      onFocus={onFocus}
      onBlur={onBlur}
    />)}
  </div>;
};

export default styled(Wrapper)`
  position: relative;
  overflow: visible;
  z-index: ${theme.zIndex.highlightInlineCard};
  top: 0;
  transition: all 0.3s;
`;
