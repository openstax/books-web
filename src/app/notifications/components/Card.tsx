import styled from 'styled-components/macro';
import { ButtonGroup } from '../../components/Button';
import { bodyCopyRegularStyle } from '../../components/Typography';
import theme from '../../theme';
import { inlineDisplayBreak } from '../theme';

// tslint:disable-next-line:variable-name
export const Group = styled.div`
  width: 100%;

  ${ButtonGroup} {
    padding: 1rem;
    margin: 0;
  }

  @media (max-width: ${inlineDisplayBreak}) {
    > ${ButtonGroup} {
      display: block;
      padding: ${theme.padding.page.mobile}rem 0 0 0;
    }
  }
`;

// tslint:disable-next-line:variable-name
export const P = styled.p`
  ${bodyCopyRegularStyle}
  margin: 0;
  padding: 1rem;

  @media (max-width: ${inlineDisplayBreak}) {
    padding: 0;
  }
`;

// tslint:disable-next-line:variable-name
export const Body = styled.div`
  margin: 0.5rem;
  background-color: ${theme.color.neutral.base};
  border-style: solid;
  border-color: ${theme.color.neutral.darkest};
  display: flex;
  flex-basis: 100%;
  flex-direction: column;
  border-width: thin;
  box-shadow: 0 1rem 2rem 0 rgba(0, 0, 0, 0.2);
  overflow: visible;

  > ${ButtonGroup} {
    padding: 1rem;
    margin: 0;
  }

  @media (max-width: ${inlineDisplayBreak}) {
    margin: ${theme.padding.page.mobile}rem;
    box-shadow: none;
    border: none;
    flex-direction: row;
    background-color: initial;

    > ${ButtonGroup} {
      padding: 0;
    }
  }
`;

// tslint:disable-next-line:variable-name
export const Header = styled.div`
  ${bodyCopyRegularStyle}
  padding: 0.5rem 1rem;
  font-weight: bold;
  background-color: ${theme.color.neutral.darker};
  line-height: 4rem;
  margin-bottom: 1rem;

  @media (max-width: ${inlineDisplayBreak}) {
    padding: 0;
    line-height: 2rem;
    background-color: initial;
  }
`;
