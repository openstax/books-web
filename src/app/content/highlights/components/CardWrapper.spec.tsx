import React from 'react';
import renderer from 'react-test-renderer';
import createMockHighlight from '../../../../test/mocks/highlight';
import Card from './Card';
import CardWrapper from './CardWrapper';

jest.mock('./Card', () => () => null);
jest.mock('./cardUtils', () => ({
  getHighlightTopOffset: () => 0,
}));

describe('CardWrapper', () => {
  it('matches snapshot', async() => {
    const component = renderer.create(<CardWrapper
      highlights={[createMockHighlight()]}
    />);

    // wait for React.useEffect
    await renderer.act(async() => undefined);

    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('renders cards', async() => {
    const component = renderer.create(<CardWrapper
      highlights={[createMockHighlight(), createMockHighlight()]}
    />);

    // wait for React.useEffect
    await renderer.act(async() => undefined);

    expect(component.root.findAllByType(Card).length).toBe(2);
  });
});
