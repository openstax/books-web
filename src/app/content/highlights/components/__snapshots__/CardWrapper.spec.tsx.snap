// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CardWrapper matches snapshot 1`] = `
.c0 {
  position: relative;
  overflow: visible;
  z-index: 10;
  top: 0;
  -webkit-transition: all 0.3s;
  transition: all 0.3s;
}

<div
  className="c0"
/>
`;
